Fixed `no implicit conversion of String into Integer` Error when Adding/Editing Container Provider

DavidResende0 opened the pull request:
Follow up to https://github.com/ManageIQ/manageiq-ui-classic/pull/9003 which introduced the method `trimFieldValue`, that's being called on (validation and) form submit.

Before introducing `trimFieldValue` method
```
endpoints: [
    {
        "role": "default",
        "security_protocol": "ssl-without-validation",
        "hostname": "ocp-master.cfme2.lab.eng.rdu2.redhat.com",
        "port": 8443
    }
]

```
An object returned after introducing `trimFieldValue`
```
endpoints: {
    "0": {
        "role": "default",
        "security_protocol": "ssl-without-validation",
        "hostname": "ocp-master.cfme2.lab.eng.rdu2.redhat.com",
        "port": 8443
    }
}
```

When endpoints were passed through the trimFieldValue method, they converted from an array into an object, which is not in line with what the API expects. 

This PR modification involves calling the trimFieldValue method before formatting the data to prevent this issue.

jeffibm commented on app/javascript/components/provider-form/index.jsx:
```diff
@@ -176,7 +177,7 @@ const ProviderForm = ({
       const data = {
         ...rest,
-        endpoints: trimFieldValue(endpoints),
+        endpoints,
```
The method `trimFieldValue` should receive an object, modify it, and return the modified object.

But, while submitting, the method received an array from `endpoints`.

I don't remember why I added the method only for `endpoints`.

I guess the below change would work. Could you please check...

```
const data = trimFieldValue({
        ...rest,
        endpoints,
        authentications,
        ...(edit ? undefined : { type }),
        ddf: true,
      });
```

jeffibm commented:
Probably better to check for objects --

```
/** Field names that need to be trimmed. */
const trimFields = ['hostname'];

const isObject = (data) => (typeof data === 'object' && data !== null);

/** Function to trim a field's value.
 * If the nested data object contains the key which includes in variable 'trimFields',
 * then, trim the spaces from the value before sending to API.
 */
export const trimFieldValue = (resource) => {
  if (!isObject(resource)) {
    return resource;
  }
  const keys = Object.keys(resource);

  keys.find((key) => {
    const value = resource[key];

    if (trimFields.includes(key)) {
      resource[key] = resource[key] ? resource[key].trim() : '';
      return true;
    } if (isObject(value)) {
      return trimFieldValue(value) !== undefined;
    }

    return false;
  });

  return { ...resource };
};

```

jeffibm commented:
And maybe add a test case for it - 

```
describe('trimFieldValue function', () => {
  it('should trim the hostname field', () => {
    const resource = {
      hostname: '  example.com  ',
      otherField: 'value',
    };

    const trimmedResource = trimFieldValue(resource);

    expect(trimmedResource.hostname).toEqual('example.com');
    expect(trimmedResource.otherField).toEqual('value');
  });

  it('should handle nested objects and trim the hostname field', () => {
    const resource = {
      nestedObject: {
        hostname: '  nested-example.com  ',
      },
      otherField: 'value',
    };

    const trimmedResource = trimFieldValue(resource);

    expect(trimmedResource.nestedObject.hostname).toEqual('nested-example.com');
    expect(trimmedResource.otherField).toEqual('value');
  });

  it('should handle multiple nested objects and trim the hostname field', () => {
    const resource = {
      firstLevel: {
        nestedObject: {
          hostname: '  nested-example.com  ',
        },
      },
      otherField: 'value',
    };

    const trimmedResource = trimFieldValue(resource);

    expect(trimmedResource.firstLevel.nestedObject.hostname).toEqual('nested-example.com');
    expect(trimmedResource.otherField).toEqual('value');
  });

  it('should handle undefined values', () => {
    const resource = {
      hostname: undefined,
    };

    const trimmedResource = trimFieldValue(resource);

    expect(trimmedResource.hostname).toEqual('');
  });

  it('should handle null values', () => {
    const resource = {
      hostname: null,
    };

    const trimmedResource = trimFieldValue(resource);

    expect(trimmedResource.hostname).toEqual('');
  });

  it('should not modify non-trimmed fields', () => {
    const resource = {
      username: 'user123',
      email: 'user@example.com',
    };

    const trimmedResource = trimFieldValue(resource);

    expect(trimmedResource.username).toEqual('user123');
    expect(trimmedResource.email).toEqual('user@example.com');
  });

  it('should handle non-object values', () => {
    const resource = 'Hello';

    const trimmedResource = trimFieldValue(resource);

    expect(trimmedResource).toEqual('Hello');
  });

  it('should handle non-object values inside an array', () => {
    const resource = ['  example.com  ', '  nested-example.com  '];

    const trimmedResource = trimFieldValue(resource);

    expect(trimmedResource).toEqual(['example.com', 'nested-example.com']);
  });
});
```

DavidResende0 commented:
@jeffibm, @Fryguy I made all the changes you suggested and also added the specs. I also modified `trimFieldValue` so that it can handle arrays in the future and added more specs to test this.
